### ISSUE TITLE
-w gives DynaLoader::dl_error redefined warning from static load;    -Mwarnings doesn't

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1422]
Migrated from [rt.perl.org#2569](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=2569) (status was 'resolved')

Searchable as RT2569$

**Comment by p5pRT** [comment]
### From domo@computer.org

#### Created by domo@computer.org



$ perl \-w \-MIO \-e 1      \# \(Or any XS module of your choice\)
Subroutine DynaLoader&#8203;::dl\_error redefined at 
/usr/local/lib/perl5/5\.6\.0/powerpc\-machten/DynaLoader\.pm line 93\.
$ perl \-Mwarnings \-MIO \-e 1
$ perl \-Mwarnings=all \-MCPAN \-e  1
$

Firstly\, I don't think I should be getting a warning at all\.  I assume
it's something to do with perl being statically linked\.  Secondly\, I'm
surprised that the behaviour's different with C\<\-Mwarnings> instead of
C\<\-w>\.

<details><summary>Perl Info</summary>

```


Site configuration information for perl v5.6.0:

Configured by domo at Thu Mar 16 10:29:32 WET 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
   Platform:
     osname=machten, osvers=4.1.4, archname=powerpc-machten
     uname='machten ppp100 5 0.5 powerpc '
     config_args='-de -Doptimize=-g'
     hint=recommended, useposix=true, d_sigaction=define
     usethreads=undef use5005threads=undef useithreads=undef 
usemultiplicity=undef
     useperlio=undef d_sfio=undef uselargefiles=define
     use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
   Compiler:
     cc='cc', optimize='-g', gccversion=2.8.1
     cppflags='-DNOTDEF_MACHTEN -DREG_INFTY=2047 -DDEBUGGING 
-I/usr/local/include'
     ccflags ='-DNOTDEF_MACHTEN -DREG_INFTY=2047 -DDEBUGGING 
-I/usr/local/include'
     stdchar='char', d_stdstdio=undef, usevfork=true
     intsize=4, longsize=4, ptrsize=4, doublesize=8
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, 
Off_t='off_t', lseeksize=4
     alignbytes=8, usemymalloc=y, prototype=define
   Linker and Libraries:
     ld='ld', ldflags =' -Xlstack=1048576 -L/usr/local/lib'
     libpth=/usr/local/lib /usr/lib
     libs=-lndbm -lgdbm -ldb -lm -lc
     libc=/usr/lib/libc.a, so=so, useshrplib=false, libperl=libperl.a
   Dynamic Linking:
     dlsrc=dl_none.xs, dlext=none, d_dlsymun=undef, ccdlflags=''
     cccdlflags='', lddlflags=''

Locally applied patches:
     v5.6.0-RC2


@INC for perl v5.6.0:
     lib
     /usr/local/lib/perl5/5.6.0/powerpc-machten
     /usr/local/lib/perl5/5.6.0
     /usr/local/lib/perl5/site_perl/5.6.0/powerpc-machten
     /usr/local/lib/perl5/site_perl/5.6.0
     /usr/local/lib/perl5/site_perl/5.00557-hidden/powerpc-machten
     /usr/local/lib/perl5/site_perl/5.00557-hidden
     /usr/local/lib/perl5/site_perl/5.00556-hidden/powerpc-machten
     /usr/local/lib/perl5/site_perl/5.00556-hidden
     /usr/local/lib/perl5/site_perl
     .


Environment for perl v5.6.0:
     HOME=/home/domo
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH=/usr/lib
     LOGDIR (unset)
 
PATH=/sbin:/usr/sbin:/home/domo/bin:/usr/local/bin:/usr/bin:/bin:/usr/X11/bin:/usr/libexec
     PERL_BADLANG (unset)
     SHELL=/bin/bash
-- 
Dominic Dunlop


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

From&#8203;: Dominic Dunlop \[mailto&#8203;:domo@&#8203;computer\.org\]
> 
> $ perl \-w \-MIO \-e 1      \# \(Or any XS module of your choice\)
> Subroutine DynaLoader&#8203;::dl\_error redefined at 
> /usr/local/lib/perl5/5\.6\.0/powerpc\-machten/DynaLoader\.pm line 93\.
> $ perl \-Mwarnings \-MIO \-e 1
> $ perl \-Mwarnings=all \-MCPAN \-e  1
> $
> 
> Firstly\, I don't think I should be getting a warning at all\.  I assume
> it's something to do with perl being statically linked\.  Secondly\, I'm
> surprised that the behaviour's different with C\<\-Mwarnings> instead of
> C\<\-w>\.

The issue of the difference between \-Mwarnings and \-w is probably because
the "warnings" module is lexical\. Hence

&nbsp;    perl \-Mwarnings \-MIO \-e 1

means roughly

&nbsp;    perl \<\<EOF
&nbsp;        use warnings;
&nbsp;        use IO;
&nbsp;        1;
&nbsp;    EOF

which doesn't propogate the warning setting through to the IO module\. \-w\, on
the other hand\, affects \*all\* included code\.

This is probably worth a note in the documentation\. I'm not 100% sure which
should be recommended\, but the difference is important\.

Paul\.

**Comment by p5pRT** [comment]
### From @pmqs

From&#8203;: Moore\, Paul \[mailto&#8203;:Paul\.Moore@&#8203;uk\.origin\-it\.com\]

> From&#8203;: Dominic Dunlop \[mailto&#8203;:domo@&#8203;computer\.org\]
> > 
> > $ perl \-w \-MIO \-e 1      \# \(Or any XS module of your choice\)
> > Subroutine DynaLoader&#8203;::dl\_error redefined at 
> > /usr/local/lib/perl5/5\.6\.0/powerpc\-machten/DynaLoader\.pm line 93\.
> > $ perl \-Mwarnings \-MIO \-e 1
> > $ perl \-Mwarnings=all \-MCPAN \-e  1
> > $
> > 
> > Firstly\, I don't think I should be getting a warning at all\.  I assume
> > it's something to do with perl being statically linked\.  Secondly\, I'm
> > surprised that the behaviour's different with C\<\-Mwarnings> instead of
> > C\<\-w>\.
> 
> The issue of the difference between \-Mwarnings and \-w is probably because
> the "warnings" module is lexical\. Hence
> 
>     perl \-Mwarnings \-MIO \-e 1
> 
> means roughly
> 
>     perl \<\<EOF
>         use warnings;
>         use IO;
>         1;
>     EOF
> 
> which doesn't propogate the warning setting through to the IO module\. \-w\,
on
> the other hand\, affects \*all\* included code\.

Got it in one\.

Paul

**Comment by p5pRT** [comment]
### From @gsar

On Wed\, 22 Mar 2000 12&#8203;:28&#8203;:14 \+0100\, Dominic Dunlop wrote&#8203;:
>$ perl \-w \-MIO \-e 1      \# \(Or any XS module of your choice\)

I don't think the bit about any XS module is true\.  It only happens
for IO here\.

>Subroutine DynaLoader&#8203;::dl\_error redefined at 
>/usr/local/lib/perl5/5\.6\.0/powerpc\-machten/DynaLoader\.pm line 93\.

Patch attached\.


Sarathy
gsar@&#8203;ActiveState\.com
<details><summary>Inline Patch</summary>

```diff
-----------------------------------8<-----------------------------------
Change 5884 by gsar@auger on 2000/03/22 18:16:45

	avoid loading both XSLoader and DynaLoader (avoids dl_error()
	redefinition warnings when statically linked)

Affected files ...

... //depot/perl/ext/IO/lib/IO/File.pm#11 edit

Differences ...

==== //depot/perl/ext/IO/lib/IO/File.pm#11 (text) ====
Index: perl/ext/IO/lib/IO/File.pm
--- perl/ext/IO/lib/IO/File.pm.~1~	Wed Mar 22 10:16:50 2000
+++ perl/ext/IO/lib/IO/File.pm	Wed Mar 22 10:16:50 2000
@@ -113,9 +113,8 @@
 use File::Spec;
 
 require Exporter;
-require DynaLoader;
 
-@ISA = qw(IO::Handle IO::Seekable Exporter DynaLoader);
+@ISA = qw(IO::Handle IO::Seekable Exporter);
 
 $VERSION = "1.08";
 
End of Patch.
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

At 10&#8203;:18 \-0800 2000\-03\-22\, Gurusamy Sarathy wrote&#8203;:
>I don't think the bit about any XS module is true\.  It only happens
>for IO here\.

You're right\.  The most cursory examination shows that the few I 
happened to try all used IO\.  Apologies\.

>Patch attached\.

Yup\, that does it\.  Thank you\.
\-\- 
Dominic Dunlop

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

This ticket can be closed&#8203;: patch 5947 and/or 7269 \(which appear to be 
the same patch on different branches\.  Or something\) fixed it\, along 
with two others\.
